### PR TITLE
Add ability to handle streams returned from controller

### LINKED ISSUE #551: Ability to handle streams returned from controller
I have a use-case where I want my controller to pipe streams through to the response object (eg. proxying from s3). This isn't possible to be encoded currently, but the changes required seem pretty minimal. 

## Sorting

* **I'm submitting a ...**
  - [x] feature request

* I confirm that I
  - [x] used the [search](https://github.com/lukeautry/tsoa/search?type=Issues) to make sure that a similar issue hasn't already been submit

## Expected Behavior

- Controllers can return a `Readable` stream object instead of a `Buffer`
- The swagger spec should be the same as if you returned a buffer directly

## Possible Solution

- Add check to `promiseHandler` in route templates that pipes the response rather than sending it via `json`
- Swagger generator should be aware of the `Readable` type and output the same return type as a `Buffer`

## Context (Environment)
Version of the library: 2.5.10
Version of NodeJS: 12
* Confirm you were using yarn not npm: [x]

## Breaking change?

Not a breaking change.

### PR DESCRIPTION
**Pull request by jtfell**
I have a use-case where I want my controller to pipe streams through to the response object (eg. proxying from s3). This isn't possible to be encoded currently, but the changes required seem pretty minimal. 

My assumptions:

- The swagger spec should be the same as if you returned a buffer directly.
- The route template should pipe the response rather than sending it via `json`

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing](https://github.com/lukeautry/tsoa/tree/master/docs/CONTRIBUTING.md) document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/lukeautry/tsoa/pulls) for the same update/change?
* [x] Have you written unit tests?
* [x] Have you written unit tests that cover the negative cases (i.e.: if bad data is submitted, does the library respond properly)?
* [x] This PR is associated with an existing issue?

Resolves #551

### If this is a new feature submission:

* [x] Has the issue had a maintainer respond to the issue and clarify that the feature is something that aligns with the [goals](https://github.com/lukeautry/tsoa#goal) and [philosophy](https://github.com/lukeautry/tsoa#philosophy) of the project?

**Potential Problems With The Approach**

- I couldn't find any relevant tests for controllers using buffers, as the structure here should be quite similar I imagine
- I'm not experienced with Koa or Hapi so wasn't sure how to integrate the `data.pipe(res)` pattern that I used in the express routes template

**Test plan**

Some guidance here would be great.

### CODE REVIEW

File: src/routeGeneration/templates/express.hbs [hunk 1]
```diff
@@ -161,7 +161,9 @@ export function RegisterRoutes(app: express.Express) {
 
                 // WARNING: This file was auto-generated with tsoa. Please do not modify it. Re-run tsoa to re-generate this file: https://github.com/lukeautry/tsoa
 
-                if (data || data === false) { // === false allows boolean result
+                if (data && data.readable) {
```
Comment (WoH) [inline src/routeGeneration/templates/express.hbs]: This is a dangerous check, please make sure this only runs when data is an instance of `Readable`.

Also, please copy this to the other templates.

File: src/metadataGeneration/tsoa.ts [hunk 2]
```diff
@@ -182,6 +184,10 @@ export namespace Tsoa {
     dataType: 'buffer';
   }
 
+  export interface StreamType extends TypeBase {
```
Comment (WoH) [inline src/metadataGeneration/tsoa.ts]: You can avoid adding a new type (and therefore, a lot of code) if you reuse `BufferType`.

File: src/routeGeneration/templates/express.hbs (hunk 1 above)
Comment (jtfell) [reply src/routeGeneration/templates/express.hbs]: Are you suggesting a guard like: `data && data instanceof Readable`? This is not the recommended way according to this SO (from 2014) https://stackoverflow.com/questions/23885095/nodejs-check-if-variable-is-readable-stream. They suggest using duck-typing so something more like:

`data && data.pipe && data._read`

Alternatively [this package](https://github.com/sindresorhus/is-stream) is an option.

I'm open to whichever, what do you prefer? @WoH

File: src/metadataGeneration/tsoa.ts (hunk 2 above)
Comment (jtfell) [reply src/metadataGeneration/tsoa.ts]: Good idea. I've pushed a new commit to reuse `BufferType`.

**Comment by WoH** [comment]
I'm not sure we want to handle legacy streams or arbitrarily call pipe on anything that responds to that method, but there's no good solution here that I'm aware of.
I'd accept a check that checks for a method called `pipe` and makes sure the stream is readable.

File: src/routeGeneration/templates/express.hbs (hunk 1 above)
Comment (WoH) [reply src/routeGeneration/templates/express.hbs]: Good points. I didn't consider handling anything except the Node ReadableStream. I'm not sure we want to handle legacy streams or arbitrarily call pipe on anything that responds to that method, but there's no perfect solution here that I'm aware of.
I'd accept a check that checks for a function called pipe and makes sure the stream is readable.

Comment (jtfell) [reply src/routeGeneration/templates/express.hbs]: I've gone with this check: `data && typeof data.pipe === 'function' && data.readable && typeof data._read === 'function'`.

It looks like both Hapi and Koa will automatically handle streams so no further changes should be needed for the other templates.

**Comment by WoH** [comment]
Is there a way to make this work across all supported frameworks?

**Comment by jtfell** [comment]
It should already @WoH - Koa and Hapi both support streams without changes to the templates.

**Pull request merged**